src/pgcopy/src/copy.rs

```diff
@@ -495,8 +508,60 @@ pub fn decode_copy_format_csv(
                 row.push(Datum::Null);
             } else {
                 match mz_pgrepr::Value::decode_text(typ, raw_value) {
-                    Ok(value) => row.push(value.into_datum(&buf, typ)),
+                    Ok(mut value) => {
+                        // From
+                        // https://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-data-conversion.html#copy-truncatecolumns:
+                        //
+                        // Truncates data in VARCHAR and CHAR columns to the appropriate number of
+                        // characters so that it fits the column specification.
+                        if truncate_columns {
+                            match (typ, &mut value) {
+                                (
+                                    Type::VarChar {
+                                        max_length: Some(max_length),
+                                    },
+                                    Value::VarChar(v),
+                                ) => {
+                                    // Count utf8 chars, then truncate if needed. Fairly slow due to
+                                    // utf8 math.
+                                    let max_length: usize = match max_length.into_i32().try_into() {
+                                        Ok(v) if v > 0 => v,
+                                        _ => usize::MAX,
+                                    };
+                                    let count = v.chars().count();
+                                    if count > max_length {
+                                        *v = v.chars().take(max_length).collect();
+                                    }
```

Comment:
If `count` is slow, it might be worth replacing all 4 lines with just `*v = v.chars().take(max_length).collect();`. That way we avoid counting twice and avoid ever counting more characters than `max_length`. 

I believe that in the case of `count <= max_length` it has no effect on v.